dirdiff

```diff
@@ -0,0 +1,275 @@
+    )
+    parser.add_argument(
+        "--ignore-moved",
+        action="store_true",
+        help="don't show files that were moved around",
+    )
+    parser.add_argument(
+        "--ignore-additional",
+        action="store_true",
+        help="don't show files if they weren't either moved or modified",
+    )
+    parser.add_argument(
+        "--json",
+        action="store_true",
+        help="display output as JSON",
+    )
+    parser.add_argument(
+        "dirpath",
+        type=Path,
+        nargs="+",
+        help="dir path",
+    )
+    args = parser.parse_args()
+    if args.debug:
+        LOG.setLevel(logging.DEBUG)
+    return args
+
+
+@contextmanager
+def chdir(dirname):
+    oldpwd = Path().absolute()
+    try:
+        os.chdir(dirname)
+        yield
+    finally:
+        os.chdir(oldpwd)
+
+
+def apply_to_dir(dirpath, func):
+    return [
+        func(os.path.join(dirp, fname))
+        for dirp, _, filenames in os.walk(dirpath)
+        for fname in filenames
+    ]
+
+
+def file_hexdigest(filepath):
+    hasher = hashlib.sha1()
+    BUFSIZ = 65536
+    try:
+        with open(filepath, "rb") as f:
+            buf = f.read(BUFSIZ)
+            while len(buf) > 0:
+                hasher.update(buf)
+                buf = f.read(BUFSIZ)
+    except FileNotFoundError as e:
+        # if this is a symlink
+        if stat.S_ISLNK(os.lstat(filepath).st_mode):
+            hasher.update(b"")
+        else:
+            raise
+    return hasher.hexdigest()
+
+
+def preprocess_into_datastructs(rootdirs: List[Path]):
+    # let's collect all the files from all the rootdirs and build our base data structures
+    hashdiff_data: Dict[HexDigest, Dict[Path, List[DirDiffFile]]] = defaultdict(
+        lambda: defaultdict(list)
+    )
+    pathdiff_data: Dict[Path, List[DirDiffFile]] = defaultdict(list)
+
+    for rootdir in rootdirs:
+        rootdir_path = Path(rootdir)
+        with chdir(rootdir_path):
+            filelist = apply_to_dir(".", lambda x: Path(x))
+        for f in filelist:
+            try:
+                ddf = DirDiffFile(rootdir_path, f)
+                hashdiff_data[ddf.hexdigest][ddf.position].append(ddf)
+                pathdiff_data[ddf.position].append(ddf)
+            except PermissionError as e:
+                LOG.warning(f"error processing {f}: {e}")
+
+    return hashdiff_data, pathdiff_data
+
+
+def partition_matched_unmatched(
+    hashdiff_data: Dict[HexDigest, Dict[Path, List[DirDiffFile]]], rootdirs: List[Path]
+):
+    matched: Dict[HexDigest, List[DirDiffFile]] = defaultdict(list)
+    unmatched: Dict[HexDigest, List[DirDiffFile]] = defaultdict(list)
+
+    # dir1/foopath1/barfile1 is matched with dir2/foopath2/barfile2 if:
+    #    - foopath1 == foopath2
+    #    - hash(barfile1) == hash(barfile2).
+    # Otherwise, barfile1 is unmatched.
+    len_rootdirs = len(rootdirs)
+    for hexdigest, filedata in hashdiff_data.items():
+        for _, filelist in filedata.items():
```

Comment:
simplification: `for filelist in filedata.values()`

Reply:
Personal taste time: I still prefer `dict.items()` because if using a dict, I'm interested in the key as much as the value. For future-proofing reasons, I use `.items()` in case I'll need to refer to the key in the future.